File: ios/SecretaryKit/Sources/SecretaryKit/VaultAccess/MonitorSyncHook.swift
import Foundation
import SecretaryVaultAccess

/// Real `SyncMonitorHook` over a `ChangeDetectionMonitor`: mute around our own
/// vault writes (a window starting now) and acknowledge handled changes.
@MainActor
public final class MonitorSyncHook: SyncMonitorHook {
    private let monitor: ChangeDetectionMonitor
    private let muteWindow: Duration
    public init(monitor: ChangeDetectionMonitor,
                muteWindow: Duration = ChangeDetectionTuning.defaultSelfWriteMuteWindow) {
        self.monitor = monitor
        self.muteWindow = muteWindow
    }
    public func muteSelfWrite() {
        monitor.muteUntil(MonotonicInstant.now().advanced(by: muteWindow))
    }
    public func acknowledge() {
        monitor.acknowledge()
    }
}

File: android/kit/src/main/kotlin/org/secretary/sync/MonitorSyncHook.kt
package org.secretary.sync

import kotlin.time.Duration

/**
 * Real [SyncMonitorHook] wrapping a [ChangeDetectionMonitor]. [muteSelfWrite] suppresses detector
 * pulses for [muteWindow] from now (so a sync pass's own manifest rewrite is not re-detected);
 * [acknowledge] consumes the monitor's pending-change signal after a clean pass. [now] defaults to
 * the Android monotonic clock and is injected only for host tests.
 */
class MonitorSyncHook(
    private val monitor: ChangeDetectionMonitor,
    private val muteWindow: Duration = ChangeDetectionTuning.defaultSelfWriteMuteWindow,
    private val now: () -> MonotonicInstant = ::monotonicNow,
) : SyncMonitorHook {
    override fun muteSelfWrite() {
        monitor.muteUntil(now().advancedBy(muteWindow))
    }

    override fun acknowledge() {
        monitor.acknowledge()
    }
}
